src/main/java/com/faforever/client/tournament/game/TournamentGameService.java

```diff
@@ -0,0 +1,122 @@
+    fafServerAccessor.addEventListener(IsReadyRequest.class, this::onReadRequest);
+    fafServerAccessor.addEventListener(MatchmakerMatchCancelledResponse.class, this::onMatchCanceled);
+    fafServerAccessor.addEventListener(GameLaunchResponse.class, this::onGameLaunch);
+  }
+
+  private void onGameLaunch(GameLaunchResponse gameLaunchResponse) {
+    dismissNotification();
+  }
+
+  private void onMatchCanceled(MatchmakerMatchCancelledResponse matchmakerMatchCancelledResponse) {
+    dismissNotification();
+    cancelMatch();
+    notificationService.addImmediateInfoNotification(
+        "isReady.matchCanceled"
+    );
+  }
+
+  private void onReadRequest(IsReadyRequest isReadyRequest) {
+    log.info("Tournament game is ready, asking user.");
+    if (notification != null) {
+      log.warn("Tournament ready request ignored because tournament is already in progress.");
+      respondToReadyRequest(isReadyRequest.getRequestId(), isReadyRequest);
+      return;
+    }
+    uiService.bringMainStageToFront();
+    final var controller = initializeIsReadyController(isReadyRequest);
+    notification =
+        new ImmediateNotification(i18n.get("isReady.title"), i18n.get("isReady.message", isReadyRequest.getGameName()),
+            Severity.INFO, null, List.of(), controller.getRoot(), false);
+    notificationService.addNotification(notification);
+    safetyCloseMechanism();
+  }
+
+  private void safetyCloseMechanism() {
+    final var currentNotification = notification;
+    timer.schedule(new TimerTask() {
+      @Override
+      public void run() {
+        if (notification != currentNotification) {
+          return;
+        }
+        JavaFxUtil.runLater(() -> dismissNotification());
+      }
+    }, SAFETY_CLOSE_NOTIFICATION_SECONDS * 1000);
+  }
+
+  private void dismissNotification() {
+    if (notification == null) {
+      return;
+    }
+    notification.dismiss();
+    notification = null;
+  }
+
+  @NotNull
+  private IsReadyForGameController initializeIsReadyController(IsReadyRequest isReadyRequest) {
+    IsReadyForGameController controller = uiService.loadFxml("theme/tournaments/is_ready_for_game.fxml");
+    controller.setTimeout(isReadyRequest.getResponseTimeSeconds());
+    controller.setReadyCallback(() -> respondToReadyRequest(isReadyRequest.getRequestId(), isReadyRequest));
+    controller.setDismissCallBack(() -> JavaFxUtil.runLater(this::dismissNotification));
+    return controller;
+  }
+
+
+  private void respondToReadyRequest(String requestId, IsReadyRequest isReadyRequest) {
+    matchFuture = gameService.startListeningToTournamentGame(isReadyRequest.getFeaturedMod());
+    try {
+      fafServerAccessor.sendIsReady(requestId);
+    } catch (Exception e) {
+      dismissNotification();
+      cancelMatch();
+      notificationService.addImmediateErrorNotification(e, "isReady.readyUpFailed");
+    }
+  }
+
+  private void cancelMatch() {
+    if (matchFuture != null) {
+      matchFuture.cancel(false);
```

Comment:
Just a technical note the match future is not interruptible so only non started stages will be cancelled. This means that the game launch can still happen in some cases. Active problem with matchmaker cancellation as well

Reply:
Hmm ok. Yes mechanism is the same as for tmm. Would be good to fix here as well should it ever be fixed.

Reply:
Yeah it is mostly caused by the CompletableFutures not having an interrupt mechanism so would have to go more low level to fix it in a nice way